// @liveblocks/react-lexical/dist/index.js
'use strict';

var core = require('@liveblocks/core');
var version = require('./version.js');
var anchoredThreads = require('./comments/anchored-threads.js');
var commentPluginProvider = require('./comments/comment-plugin-provider.js');
var floatingComposer = require('./comments/floating-composer.js');
var floatingThreads = require('./comments/floating-threads.js');
var liveblocksConfig = require('./liveblocks-config.js');
var liveblocksPluginProvider = require('./liveblocks-plugin-provider.js');
var historyVersionPreview = require('./version-history/history-version-preview.js');

core.detectDupes(version.PKG_NAME, version.PKG_VERSION, version.PKG_FORMAT);

exports.AnchoredThreads = anchoredThreads.AnchoredThreads;
exports.useIsThreadActive = commentPluginProvider.useIsThreadActive;
exports.FloatingComposer = floatingComposer.FloatingComposer;
exports.OPEN_FLOATING_COMPOSER_COMMAND = floatingComposer.OPEN_FLOATING_COMPOSER_COMMAND;
exports.FloatingThreads = floatingThreads.FloatingThreads;
exports.liveblocksConfig = liveblocksConfig.liveblocksConfig;
exports.LiveblocksPlugin = liveblocksPluginProvider.LiveblocksPlugin;
exports.useEditorStatus = liveblocksPluginProvider.useEditorStatus;
exports.HistoryVersionPreview = historyVersionPreview.HistoryVersionPreview;
