// @liveblocks/react-ui/dist/components/Comment.js
'use client';
'use strict';

var react = require('@liveblocks/react');
var TogglePrimitive = require('@radix-ui/react-toggle');
var React = require('react');
var Check = require('../icons/Check.js');
var Cross = require('../icons/Cross.js');
var Delete = require('../icons/Delete.js');
var Edit = require('../icons/Edit.js');
var Ellipsis = require('../icons/Ellipsis.js');
var EmojiAdd = require('../icons/EmojiAdd.js');
var overrides = require('../overrides.js');
var index$1 = require('../primitives/Comment/index.js');
var index = require('../primitives/Composer/index.js');
var Timestamp = require('../primitives/Timestamp.js');
var shared = require('../shared.js');
var mentions = require('../slate/plugins/mentions.js');
var classNames = require('../utils/class-names.js');
var useRefs = require('../utils/use-refs.js');
var useVisible = require('../utils/use-visible.js');
var useWindowFocus = require('../utils/use-window-focus.js');
var Composer = require('./Composer.js');
var Avatar = require('./internal/Avatar.js');
var Button = require('./internal/Button.js');
var Dropdown = require('./internal/Dropdown.js');
var Emoji = require('./internal/Emoji.js');
var EmojiPicker = require('./internal/EmojiPicker.js');
var List = require('./internal/List.js');
var Tooltip = require('./internal/Tooltip.js');
var User = require('./internal/User.js');
var TooltipPrimitive = require('@radix-ui/react-tooltip');
var PopoverPrimitive = require('@radix-ui/react-popover');
var DropdownMenuPrimitive = require('@radix-ui/react-dropdown-menu');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var TogglePrimitive__namespace = /*#__PURE__*/_interopNamespaceDefault(TogglePrimitive);

const REACTIONS_TRUNCATE = 5;
function CommentMention({
  userId,
  className,
  ...props
}) {
  const currentId = shared.useCurrentUserId();
  return /* @__PURE__ */ React.createElement(index$1.Mention, {
    className: classNames.classNames("lb-comment-mention", className),
    "data-self": userId === currentId ? "" : void 0,
    ...props
  }, mentions.MENTION_CHARACTER, /* @__PURE__ */ React.createElement(User.User, {
    userId
  }));
}
function CommentLink({
  href,
  children,
  className,
  ...props
}) {
  return /* @__PURE__ */ React.createElement(index$1.Link, {
    className: classNames.classNames("lb-comment-link", className),
    href,
    ...props
  }, children);
}
function CommentNonInteractiveLink({
  href: _href,
  children,
  className,
  ...props
}) {
  return /* @__PURE__ */ React.createElement("span", {
    className: classNames.classNames("lb-comment-link", className),
    ...props
  }, children);
}
const CommentReactionButton = React.forwardRef(({ reaction, overrides: overrides$1, className, ...props }, forwardedRef) => {
  const $ = overrides.useOverrides(overrides$1);
  return /* @__PURE__ */ React.createElement(Button.Button, {
    className: classNames.classNames("lb-comment-reaction", className),
    variant: "outline",
    "aria-label": $.COMMENT_REACTION_DESCRIPTION(
      reaction.emoji,
      reaction.users.length
    ),
    ...props,
    ref: forwardedRef
  }, /* @__PURE__ */ React.createElement(Emoji.Emoji, {
    className: "lb-comment-reaction-emoji",
    emoji: reaction.emoji
  }), /* @__PURE__ */ React.createElement("span", {
    className: "lb-comment-reaction-count"
  }, reaction.users.length));
});
const CommentReaction = React.forwardRef(({ comment, reaction, overrides: overrides$1, disabled, ...props }, forwardedRef) => {
  const addReaction = react.useAddReaction();
  const removeReaction = react.useRemoveReaction();
  const currentId = shared.useCurrentUserId();
  const isActive = React.useMemo(() => {
    return reaction.users.some((users) => users.id === currentId);
  }, [currentId, reaction]);
  const $ = overrides.useOverrides(overrides$1);
  const tooltipContent = React.useMemo(
    () => /* @__PURE__ */ React.createElement("span", null, $.COMMENT_REACTION_LIST(
      /* @__PURE__ */ React.createElement(List.List, {
        values: reaction.users.map((users) => /* @__PURE__ */ React.createElement(User.User, {
          key: users.id,
          userId: users.id,
          replaceSelf: true
        })),
        formatRemaining: $.LIST_REMAINING_USERS,
        truncate: REACTIONS_TRUNCATE,
        locale: $.locale
      }),
      reaction.emoji,
      reaction.users.length
    )),
    [$, reaction]
  );
  const stopPropagation = React.useCallback((event) => {
    event.stopPropagation();
  }, []);
  const handlePressedChange = React.useCallback(
    (isPressed) => {
      if (isPressed) {
        addReaction({
          threadId: comment.threadId,
          commentId: comment.id,
          emoji: reaction.emoji
        });
      } else {
        removeReaction({
          threadId: comment.threadId,
          commentId: comment.id,
          emoji: reaction.emoji
        });
      }
    },
    [addReaction, comment.threadId, comment.id, reaction.emoji, removeReaction]
  );
  return /* @__PURE__ */ React.createElement(Tooltip.Tooltip, {
    content: tooltipContent,
    multiline: true,
    className: "lb-comment-reaction-tooltip"
  }, /* @__PURE__ */ React.createElement(TogglePrimitive__namespace.Root, {
    asChild: true,
    pressed: isActive,
    onPressedChange: handlePressedChange,
    onClick: stopPropagation,
    disabled,
    ref: forwardedRef
  }, /* @__PURE__ */ React.createElement(CommentReactionButton, {
    "data-self": isActive ? "" : void 0,
    reaction,
    overrides: overrides$1,
    ...props
  })));
});
const CommentNonInteractiveReaction = React.forwardRef(({ reaction, overrides, ...props }, forwardedRef) => {
  const currentId = shared.useCurrentUserId();
  const isActive = React.useMemo(() => {
    return reaction.users.some((users) => users.id === currentId);
  }, [currentId, reaction]);
  return /* @__PURE__ */ React.createElement(CommentReactionButton, {
    disableable: false,
    "data-self": isActive ? "" : void 0,
    reaction,
    overrides,
    ...props,
    ref: forwardedRef
  });
});
function AutoMarkReadThreadIdHandler({
  threadId,
  commentRef
}) {
  const markThreadAsRead = react.useMarkThreadAsRead();
  const isWindowFocused = useWindowFocus.useWindowFocus();
  useVisible.useVisibleCallback(
    commentRef,
    () => {
      markThreadAsRead(threadId);
    },
    {
      enabled: isWindowFocused
    }
  );
  return null;
}
const Comment = React.forwardRef(
  ({
    comment,
    indentContent = true,
    showDeleted,
    showActions = "hover",
    showReactions = true,
    onAuthorClick,
    onMentionClick,
    onCommentEdit,
    onCommentDelete,
    overrides: overrides$1,
    className,
    additionalActions,
    additionalActionsClassName,
    autoMarkReadThreadId,
    ...props
  }, forwardedRef) => {
    const isInRoom = Boolean(React.useContext(react.RoomContext));
    const ref = React.useRef(null);
    const mergedRefs = useRefs.useRefs(forwardedRef, ref);
    const currentUserId = shared.useCurrentUserId();
    const deleteComment = react.useDeleteComment();
    const editComment = react.useEditComment();
    const addReaction = react.useAddReaction();
    const removeReaction = react.useRemoveReaction();
    const $ = overrides.useOverrides(overrides$1);
    const [isEditing, setEditing] = React.useState(false);
    const [isTarget, setTarget] = React.useState(false);
    const [isMoreActionOpen, setMoreActionOpen] = React.useState(false);
    const [isReactionActionOpen, setReactionActionOpen] = React.useState(false);
    const stopPropagation = React.useCallback((event) => {
      event.stopPropagation();
    }, []);
    const handleEdit = React.useCallback(() => {
      setEditing(true);
    }, []);
    const handleEditCancel = React.useCallback(
      (event) => {
        event.stopPropagation();
        setEditing(false);
      },
      []
    );
    const handleEditSubmit = React.useCallback(
      ({ body }, event) => {
        onCommentEdit?.(comment);
        event.preventDefault();
        setEditing(false);
        editComment({
          commentId: comment.id,
          threadId: comment.threadId,
          body
        });
      },
      [comment, editComment, onCommentEdit]
    );
    const handleDelete = React.useCallback(() => {
      onCommentDelete?.(comment);
      deleteComment({
        commentId: comment.id,
        threadId: comment.threadId
      });
    }, [comment, deleteComment, onCommentDelete]);
    const handleAuthorClick = React.useCallback(
      (event) => {
        onAuthorClick?.(comment.userId, event);
      },
      [comment.userId, onAuthorClick]
    );
    const handleReactionSelect = React.useCallback(
      (emoji) => {
        const reactionIndex = comment.reactions.findIndex(
          (reaction) => reaction.emoji === emoji
        );
        if (reactionIndex >= 0 && currentUserId && comment.reactions[reactionIndex].users.some(
          (user) => user.id === currentUserId
        )) {
          removeReaction({
            threadId: comment.threadId,
            commentId: comment.id,
            emoji
          });
        } else {
          addReaction({
            threadId: comment.threadId,
            commentId: comment.id,
            emoji
          });
        }
      },
      [
        addReaction,
        comment.id,
        comment.reactions,
        comment.threadId,
        removeReaction,
        currentUserId
      ]
    );
    React.useEffect(() => {
      const isWindowDefined = typeof window !== "undefined";
      if (!isWindowDefined)
        return;
      const hash = window.location.hash;
      const commentId = hash.slice(1);
      if (commentId === comment.id) {
        setTarget(true);
      }
    }, []);
    if (!showDeleted && !comment.body) {
      return null;
    }
    return /* @__PURE__ */ React.createElement(TooltipPrimitive.TooltipProvider, null, isInRoom && autoMarkReadThreadId && /* @__PURE__ */ React.createElement(AutoMarkReadThreadIdHandler, {
      commentRef: ref,
      threadId: autoMarkReadThreadId
    }), /* @__PURE__ */ React.createElement("div", {
      id: comment.id,
      className: classNames.classNames(
        "lb-root lb-comment",
        indentContent && "lb-comment:indent-content",
        showActions === "hover" && "lb-comment:show-actions-hover",
        (isMoreActionOpen || isReactionActionOpen) && "lb-comment:action-open",
        className
      ),
      "data-deleted": !comment.body ? "" : void 0,
      "data-editing": isEditing ? "" : void 0,
      "data-target": isTarget ? "" : void 0,
      dir: $.dir,
      ...props,
      ref: mergedRefs
    }, /* @__PURE__ */ React.createElement("div", {
      className: "lb-comment-header"
    }, /* @__PURE__ */ React.createElement("div", {
      className: "lb-comment-details"
    }, /* @__PURE__ */ React.createElement(Avatar.Avatar, {
      className: "lb-comment-avatar",
      userId: comment.userId,
      onClick: handleAuthorClick
    }), /* @__PURE__ */ React.createElement("span", {
      className: "lb-comment-details-labels"
    }, /* @__PURE__ */ React.createElement(User.User, {
      className: "lb-comment-author",
      userId: comment.userId,
      onClick: handleAuthorClick
    }), /* @__PURE__ */ React.createElement("span", {
      className: "lb-comment-date"
    }, /* @__PURE__ */ React.createElement(Timestamp.Timestamp, {
      locale: $.locale,
      date: comment.createdAt,
      className: "lb-date lb-comment-date-created"
    }), comment.editedAt && comment.body && /* @__PURE__ */ React.createElement(React.Fragment, null, " ", /* @__PURE__ */ React.createElement("span", {
      className: "lb-comment-date-edited"
    }, $.COMMENT_EDITED))))), showActions && !isEditing && /* @__PURE__ */ React.createElement("div", {
      className: classNames.classNames(
        "lb-comment-actions",
        additionalActionsClassName
      )
    }, additionalActions ?? null, showReactions && /* @__PURE__ */ React.createElement(EmojiPicker.EmojiPicker, {
      onEmojiSelect: handleReactionSelect,
      onOpenChange: setReactionActionOpen
    }, /* @__PURE__ */ React.createElement(Tooltip.Tooltip, {
      content: $.COMMENT_ADD_REACTION
    }, /* @__PURE__ */ React.createElement(PopoverPrimitive.PopoverTrigger, {
      asChild: true
    }, /* @__PURE__ */ React.createElement(Button.Button, {
      className: "lb-comment-action",
      onClick: stopPropagation,
      "aria-label": $.COMMENT_ADD_REACTION
    }, /* @__PURE__ */ React.createElement(EmojiAdd.EmojiAddIcon, {
      className: "lb-button-icon"
    }))))), comment.userId === currentUserId && /* @__PURE__ */ React.createElement(Dropdown.Dropdown, {
      open: isMoreActionOpen,
      onOpenChange: setMoreActionOpen,
      align: "end",
      content: /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(Dropdown.DropdownItem, {
        onSelect: handleEdit,
        onClick: stopPropagation
      }, /* @__PURE__ */ React.createElement(Edit.EditIcon, {
        className: "lb-dropdown-item-icon"
      }), $.COMMENT_EDIT), /* @__PURE__ */ React.createElement(Dropdown.DropdownItem, {
        onSelect: handleDelete,
        onClick: stopPropagation
      }, /* @__PURE__ */ React.createElement(Delete.DeleteIcon, {
        className: "lb-dropdown-item-icon"
      }), $.COMMENT_DELETE))
    }, /* @__PURE__ */ React.createElement(Tooltip.Tooltip, {
      content: $.COMMENT_MORE
    }, /* @__PURE__ */ React.createElement(DropdownMenuPrimitive.DropdownMenuTrigger, {
      asChild: true
    }, /* @__PURE__ */ React.createElement(Button.Button, {
      className: "lb-comment-action",
      disabled: !comment.body,
      onClick: stopPropagation,
      "aria-label": $.COMMENT_MORE
    }, /* @__PURE__ */ React.createElement(Ellipsis.EllipsisIcon, {
      className: "lb-button-icon"
    }))))))), /* @__PURE__ */ React.createElement("div", {
      className: "lb-comment-content"
    }, isEditing ? /* @__PURE__ */ React.createElement(Composer.Composer, {
      className: "lb-comment-composer",
      onComposerSubmit: handleEditSubmit,
      defaultValue: comment.body,
      autoFocus: true,
      showAttribution: false,
      actions: /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(Tooltip.Tooltip, {
        content: $.COMMENT_EDIT_COMPOSER_CANCEL,
        "aria-label": $.COMMENT_EDIT_COMPOSER_CANCEL
      }, /* @__PURE__ */ React.createElement(Button.Button, {
        className: "lb-composer-action",
        onClick: handleEditCancel
      }, /* @__PURE__ */ React.createElement(Cross.CrossIcon, {
        className: "lb-button-icon"
      }))), /* @__PURE__ */ React.createElement(Tooltip.ShortcutTooltip, {
        content: $.COMMENT_EDIT_COMPOSER_SAVE,
        shortcut: /* @__PURE__ */ React.createElement(Tooltip.ShortcutTooltipKey, {
          name: "enter"
        })
      }, /* @__PURE__ */ React.createElement(index.Submit, {
        asChild: true
      }, /* @__PURE__ */ React.createElement(Button.Button, {
        variant: "primary",
        className: "lb-composer-action",
        onClick: stopPropagation,
        "aria-label": $.COMMENT_EDIT_COMPOSER_SAVE
      }, /* @__PURE__ */ React.createElement(Check.CheckIcon, {
        className: "lb-button-icon"
      }))))),
      overrides: {
        COMPOSER_PLACEHOLDER: $.COMMENT_EDIT_COMPOSER_PLACEHOLDER
      }
    }) : comment.body ? /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(index$1.Body, {
      className: "lb-comment-body",
      body: comment.body,
      components: {
        Mention: ({ userId }) => /* @__PURE__ */ React.createElement(CommentMention, {
          userId,
          onClick: (event) => onMentionClick?.(userId, event)
        }),
        Link: CommentLink
      }
    }), showReactions && comment.reactions.length > 0 && /* @__PURE__ */ React.createElement("div", {
      className: "lb-comment-reactions"
    }, comment.reactions.map((reaction) => /* @__PURE__ */ React.createElement(CommentReaction, {
      key: reaction.emoji,
      comment,
      reaction,
      overrides: overrides$1
    })), /* @__PURE__ */ React.createElement(EmojiPicker.EmojiPicker, {
      onEmojiSelect: handleReactionSelect
    }, /* @__PURE__ */ React.createElement(Tooltip.Tooltip, {
      content: $.COMMENT_ADD_REACTION
    }, /* @__PURE__ */ React.createElement(PopoverPrimitive.PopoverTrigger, {
      asChild: true
    }, /* @__PURE__ */ React.createElement(Button.Button, {
      className: "lb-comment-reaction lb-comment-reaction-add",
      variant: "outline",
      onClick: stopPropagation,
      "aria-label": $.COMMENT_ADD_REACTION
    }, /* @__PURE__ */ React.createElement(EmojiAdd.EmojiAddIcon, {
      className: "lb-button-icon"
    }))))))) : /* @__PURE__ */ React.createElement("div", {
      className: "lb-comment-body"
    }, /* @__PURE__ */ React.createElement("p", {
      className: "lb-comment-deleted"
    }, $.COMMENT_DELETED)))));
  }
);

exports.Comment = Comment;
exports.CommentLink = CommentLink;
exports.CommentMention = CommentMention;
exports.CommentNonInteractiveLink = CommentNonInteractiveLink;
exports.CommentNonInteractiveReaction = CommentNonInteractiveReaction;
exports.CommentReaction = CommentReaction;
