// @liveblocks/react-ui/dist/components/HistoryVersionSummary.mjs
'use client';
import React__default, { forwardRef } from 'react';
import { useOverrides } from '../overrides.mjs';
import '../primitives/Comment/index.mjs';
import '../primitives/Composer/index.mjs';
import '../primitives/Composer/contexts.mjs';
import '../primitives/EmojiPicker/index.mjs';
import { Timestamp } from '../primitives/Timestamp.mjs';
import { classNames } from '../utils/class-names.mjs';
import { List } from './internal/List.mjs';
import { User } from './internal/User.mjs';

const AUTHORS_TRUNCATE = 3;
const HistoryVersionSummary = forwardRef(({ version, selected, className, ...props }, forwardedRef) => {
  const $ = useOverrides();
  return /* @__PURE__ */ React__default.createElement("button", {
    ...props,
    className: classNames("lb-root lb-history-version-summary", className),
    ref: forwardedRef,
    "data-selected": selected ? "" : void 0
  }, /* @__PURE__ */ React__default.createElement(Timestamp, {
    locale: $.locale,
    date: version.createdAt,
    className: "lb-date lb-history-version-summary-date"
  }), /* @__PURE__ */ React__default.createElement("span", {
    className: "lb-history-version-summary-authors"
  }, /* @__PURE__ */ React__default.createElement(List, {
    values: version.authors.map((author) => /* @__PURE__ */ React__default.createElement(User, {
      key: author.id,
      userId: author.id,
      replaceSelf: true
    })),
    formatRemaining: $.LIST_REMAINING_USERS,
    truncate: AUTHORS_TRUNCATE,
    locale: $.locale
  })));
});

export { HistoryVersionSummary };
