// @liveblocks/react-ui/dist/components.js
'use client';
'use strict';

var React = require('react');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

const defaultComponents = {
  Anchor: "a"
};
const ComponentsContext = React.createContext(
  void 0
);
function useComponents(components) {
  const contextComponents = React.useContext(ComponentsContext);
  return React.useMemo(
    () => ({
      ...defaultComponents,
      ...contextComponents,
      ...components
    }),
    [contextComponents, components]
  );
}
function ComponentsProvider({
  children,
  components: providerComponents
}) {
  const contextComponents = React.useContext(ComponentsContext);
  const components = React.useMemo(
    () => ({
      ...defaultComponents,
      ...contextComponents,
      ...providerComponents
    }),
    [contextComponents, providerComponents]
  );
  return /* @__PURE__ */ React__namespace.createElement(ComponentsContext.Provider, {
    value: components
  }, children);
}

exports.ComponentsContext = ComponentsContext;
exports.ComponentsProvider = ComponentsProvider;
exports.defaultComponents = defaultComponents;
exports.useComponents = useComponents;
