// @liveblocks/react-ui/dist/index.js
'use strict';

var core = require('@liveblocks/core');
var version = require('./version.js');
var Comment = require('./components/Comment.js');
var Composer = require('./components/Composer.js');
var HistoryVersionSummary = require('./components/HistoryVersionSummary.js');
var HistoryVersionSummaryList = require('./components/HistoryVersionSummaryList.js');
var InboxNotification = require('./components/InboxNotification.js');
var InboxNotificationList = require('./components/InboxNotificationList.js');
var Thread = require('./components/Thread.js');
var config = require('./config.js');
var overrides = require('./overrides.js');
var Timestamp = require('./primitives/Timestamp.js');
var shared = require('./shared.js');

core.detectDupes(version.PKG_NAME, version.PKG_VERSION, version.PKG_FORMAT);

exports.Comment = Comment.Comment;
exports.Composer = Composer.Composer;
exports.HistoryVersionSummary = HistoryVersionSummary.HistoryVersionSummary;
exports.HistoryVersionSummaryList = HistoryVersionSummaryList.HistoryVersionSummaryList;
exports.InboxNotification = InboxNotification.InboxNotification;
exports.InboxNotificationList = InboxNotificationList.InboxNotificationList;
exports.Thread = Thread.Thread;
exports.LiveblocksUIConfig = config.LiveblocksUIConfig;
exports.useOverrides = overrides.useOverrides;
exports.Timestamp = Timestamp.Timestamp;
exports.useMentionSuggestions = shared.useMentionSuggestions;
