// @liveblocks/react-ui/dist/slate/utils/get-match-range.js
'use strict';

var slate = require('slate');

const defaultOptions = {
  direction: "before"
};
function getMatchRange(editor, at, terminators = [" "], options = defaultOptions) {
  const { include, direction } = { ...defaultOptions, ...options };
  let [start, end] = slate.Range.edges(at);
  let point = start;
  function move(direction2) {
    const next = direction2 === "after" ? slate.Editor.after(editor, point, {
      unit: "character"
    }) : slate.Editor.before(editor, point, { unit: "character" });
    const nextWord = next && slate.Editor.string(
      editor,
      direction2 === "after" ? { anchor: point, focus: next } : { anchor: next, focus: point }
    );
    const lastWord = nextWord && nextWord[direction2 === "after" ? 0 : nextWord.length - 1];
    if (next && lastWord && !terminators.includes(lastWord)) {
      point = next;
      if (point.offset === 0) {
        return false;
      }
    } else {
      return false;
    }
    return true;
  }
  if (direction !== "before") {
    point = end;
    while (move("after"))
      ;
    end = point;
  }
  if (direction !== "after") {
    point = start;
    while (move("before"))
      ;
    start = point;
  }
  if (include) {
    return {
      anchor: slate.Editor.before(editor, start, { unit: "offset" }) ?? start,
      focus: slate.Editor.after(editor, end, { unit: "offset" }) ?? end
    };
  }
  return { anchor: start, focus: end };
}

exports.getMatchRange = getMatchRange;
