// @liveblocks/react/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});



































































var _chunkTNEGW2UUjs = require('./chunk-TNEGW2UU.js');

// src/index.ts
var _core = require('@liveblocks/core');
var _client = require('@liveblocks/client');
_core.detectDupes.call(void 0, _chunkTNEGW2UUjs.PKG_NAME, _chunkTNEGW2UUjs.PKG_VERSION, _chunkTNEGW2UUjs.PKG_FORMAT);


































































exports.ClientContext = _chunkTNEGW2UUjs.ClientContext; exports.ClientSideSuspense = _chunkTNEGW2UUjs.ClientSideSuspense; exports.CreateThreadError = _chunkTNEGW2UUjs.CreateThreadError; exports.LiveblocksProvider = _chunkTNEGW2UUjs.LiveblocksProvider; exports.RoomContext = _chunkTNEGW2UUjs.RoomContext; exports.RoomProvider = _chunkTNEGW2UUjs._RoomProvider; exports.createLiveblocksContext = _chunkTNEGW2UUjs.createLiveblocksContext; exports.createRoomContext = _chunkTNEGW2UUjs.createRoomContext; exports.getUmbrellaStoreForClient = _chunkTNEGW2UUjs.getUmbrellaStoreForClient; exports.selectThreads = _chunkTNEGW2UUjs.selectThreads; exports.shallow = _client.shallow; exports.useAddReaction = _chunkTNEGW2UUjs._useAddReaction; exports.useBatch = _chunkTNEGW2UUjs.useBatch; exports.useBroadcastEvent = _chunkTNEGW2UUjs._useBroadcastEvent; exports.useCanRedo = _chunkTNEGW2UUjs.useCanRedo; exports.useCanUndo = _chunkTNEGW2UUjs.useCanUndo; exports.useClient = _chunkTNEGW2UUjs.useClient; exports.useCommentsErrorListener = _chunkTNEGW2UUjs.useCommentsErrorListener; exports.useCreateComment = _chunkTNEGW2UUjs.useCreateComment; exports.useCreateThread = _chunkTNEGW2UUjs._useCreateThread; exports.useDeleteAllInboxNotifications = _chunkTNEGW2UUjs.useDeleteAllInboxNotifications; exports.useDeleteComment = _chunkTNEGW2UUjs.useDeleteComment; exports.useDeleteInboxNotification = _chunkTNEGW2UUjs.useDeleteInboxNotification; exports.useDeleteThread = _chunkTNEGW2UUjs._useDeleteThread; exports.useEditComment = _chunkTNEGW2UUjs.useEditComment; exports.useEditThreadMetadata = _chunkTNEGW2UUjs._useEditThreadMetadata; exports.useErrorListener = _chunkTNEGW2UUjs.useErrorListener; exports.useEventListener = _chunkTNEGW2UUjs._useEventListener; exports.useHistory = _chunkTNEGW2UUjs.useHistory; exports.useHistoryVersionData = _chunkTNEGW2UUjs.useHistoryVersionData; exports.useHistoryVersions = _chunkTNEGW2UUjs._useHistoryVersions; exports.useInboxNotificationThread = _chunkTNEGW2UUjs._useInboxNotificationThread; exports.useInboxNotifications = _chunkTNEGW2UUjs.useInboxNotifications; exports.useIsInsideRoom = _chunkTNEGW2UUjs._useIsInsideRoom; exports.useLostConnectionListener = _chunkTNEGW2UUjs.useLostConnectionListener; exports.useMarkAllInboxNotificationsAsRead = _chunkTNEGW2UUjs.useMarkAllInboxNotificationsAsRead; exports.useMarkInboxNotificationAsRead = _chunkTNEGW2UUjs.useMarkInboxNotificationAsRead; exports.useMarkThreadAsRead = _chunkTNEGW2UUjs.useMarkThreadAsRead; exports.useMarkThreadAsResolved = _chunkTNEGW2UUjs.useMarkThreadAsResolved; exports.useMarkThreadAsUnresolved = _chunkTNEGW2UUjs.useMarkThreadAsUnresolved; exports.useMutation = _chunkTNEGW2UUjs._useMutation; exports.useMyPresence = _chunkTNEGW2UUjs._useMyPresence; exports.useOther = _chunkTNEGW2UUjs._useOther; exports.useOthers = _chunkTNEGW2UUjs._useOthers; exports.useOthersConnectionIds = _chunkTNEGW2UUjs.useOthersConnectionIds; exports.useOthersListener = _chunkTNEGW2UUjs._useOthersListener; exports.useOthersMapped = _chunkTNEGW2UUjs._useOthersMapped; exports.useRedo = _chunkTNEGW2UUjs.useRedo; exports.useRemoveReaction = _chunkTNEGW2UUjs.useRemoveReaction; exports.useRoom = _chunkTNEGW2UUjs._useRoom; exports.useRoomInfo = _chunkTNEGW2UUjs.useRoomInfo; exports.useRoomNotificationSettings = _chunkTNEGW2UUjs.useRoomNotificationSettings; exports.useSelf = _chunkTNEGW2UUjs._useSelf; exports.useStatus = _chunkTNEGW2UUjs.useStatus; exports.useStorage = _chunkTNEGW2UUjs._useStorage; exports.useStorageRoot = _chunkTNEGW2UUjs._useStorageRoot; exports.useStorageStatus = _chunkTNEGW2UUjs.useStorageStatus; exports.useThreadSubscription = _chunkTNEGW2UUjs.useThreadSubscription; exports.useThreads = _chunkTNEGW2UUjs._useThreads; exports.useUndo = _chunkTNEGW2UUjs.useUndo; exports.useUnreadInboxNotificationsCount = _chunkTNEGW2UUjs.useUnreadInboxNotificationsCount; exports.useUpdateMyPresence = _chunkTNEGW2UUjs._useUpdateMyPresence; exports.useUpdateRoomNotificationSettings = _chunkTNEGW2UUjs.useUpdateRoomNotificationSettings; exports.useUser = _chunkTNEGW2UUjs._useUser; exports.useUserThreads_experimental = _chunkTNEGW2UUjs._useUserThreads_experimental;
