// @liveblocks/react-ui/dist/components/internal/InboxNotificationThread.js
'use strict';

var core = require('@liveblocks/core');
var React = require('react');
var overrides = require('../../overrides.js');
var index = require('../../primitives/Comment/index.js');
var classNames = require('../../utils/class-names.js');
var Comment = require('../Comment.js');
var User = require('./User.js');

const INBOX_NOTIFICATION_THREAD_MAX_COMMENTS = 3;
function InboxNotificationComment({
  comment,
  showHeader = true,
  overrides: overrides$1,
  className,
  ...props
}) {
  const $ = overrides.useOverrides(overrides$1);
  return /* @__PURE__ */ React.createElement("div", {
    className: classNames.classNames(
      "lb-root lb-inbox-notification-comment lb-comment",
      className
    ),
    ...props
  }, showHeader && /* @__PURE__ */ React.createElement("div", {
    className: "lb-comment-header"
  }, /* @__PURE__ */ React.createElement(User.User, {
    className: "lb-comment-author",
    userId: comment.userId
  })), /* @__PURE__ */ React.createElement("div", {
    className: "lb-comment-content"
  }, comment.body ? /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(index.Body, {
    className: "lb-comment-body",
    body: comment.body,
    components: {
      Mention: Comment.CommentMention,
      Link: Comment.CommentNonInteractiveLink
    }
  }), comment.reactions.length > 0 && /* @__PURE__ */ React.createElement("div", {
    className: "lb-comment-reactions"
  }, comment.reactions.map((reaction) => /* @__PURE__ */ React.createElement(Comment.CommentNonInteractiveReaction, {
    key: reaction.emoji,
    reaction,
    overrides: overrides$1,
    disabled: true
  })))) : /* @__PURE__ */ React.createElement("div", {
    className: "lb-comment-body"
  }, /* @__PURE__ */ React.createElement("p", {
    className: "lb-comment-deleted"
  }, $.COMMENT_DELETED))));
}
function findLastCommentWithMentionedId(comments, mentionedId) {
  for (let i = comments.length - 1; i >= 0; i--) {
    const comment = comments[i];
    if (comment.userId === mentionedId) {
      continue;
    }
    if (comment.body) {
      const mentionedIds = core.getMentionedIdsFromCommentBody(comment.body);
      if (mentionedIds.includes(mentionedId)) {
        return comment;
      }
    }
  }
  return;
}
function getUserIdsFromComments(comments) {
  return Array.from(new Set(comments.map((comment) => comment.userId)));
}
function generateInboxNotificationThreadContents(inboxNotification, thread, userId) {
  const unreadComments = thread.comments.filter((comment) => {
    if (!comment.body) {
      return false;
    }
    return inboxNotification.readAt ? comment.createdAt > inboxNotification.readAt && comment.createdAt <= inboxNotification.notifiedAt : comment.createdAt <= inboxNotification.notifiedAt;
  });
  if (unreadComments.length === 0) {
    const lastComments = thread.comments.filter((comment) => comment.body).slice(-INBOX_NOTIFICATION_THREAD_MAX_COMMENTS);
    return {
      type: "comments",
      unread: false,
      comments: lastComments,
      userIds: getUserIdsFromComments(lastComments),
      date: inboxNotification.notifiedAt
    };
  }
  const commentWithMention = findLastCommentWithMentionedId(
    unreadComments,
    userId
  );
  if (commentWithMention) {
    return {
      type: "mention",
      unread: true,
      comments: [commentWithMention],
      userIds: [commentWithMention.userId],
      date: commentWithMention.createdAt
    };
  }
  const lastUnreadComments = unreadComments.slice(
    -INBOX_NOTIFICATION_THREAD_MAX_COMMENTS
  );
  return {
    type: "comments",
    unread: true,
    comments: lastUnreadComments,
    userIds: getUserIdsFromComments(unreadComments),
    date: inboxNotification.notifiedAt
  };
}

exports.INBOX_NOTIFICATION_THREAD_MAX_COMMENTS = INBOX_NOTIFICATION_THREAD_MAX_COMMENTS;
exports.InboxNotificationComment = InboxNotificationComment;
exports.generateInboxNotificationThreadContents = generateInboxNotificationThreadContents;
