// @liveblocks/react-lexical/dist/mentions/avatar.mjs
import { useUser } from '@liveblocks/react';
import React__default, { forwardRef } from 'react';
import { classNames } from '../classnames.mjs';

const Avatar = forwardRef(
  function Avatar2(props, forwardedRef) {
    const { userId, className, ...spanProps } = props;
    const { user, isLoading } = useUser(userId);
    const avatar = user ? user.avatar : void 0;
    const name = user ? user.name : void 0;
    function Initials() {
      const initials = name ? getInitials(name) : void 0;
      if (initials) {
        return /* @__PURE__ */ React__default.createElement("span", {
          "aria-hidden": true,
          className: "lb-avatar-fallback"
        }, initials);
      }
      if (isLoading)
        return null;
      if (user === void 0)
        return null;
      return /* @__PURE__ */ React__default.createElement("span", {
        "aria-label": userId,
        title: userId,
        className: "lb-avatar-fallback"
      }, getInitials(userId));
    }
    return /* @__PURE__ */ React__default.createElement("span", {
      "data-loading": isLoading ? "" : void 0,
      ...spanProps,
      className: classNames("lb-avatar", className),
      ref: forwardedRef
    }, avatar && /* @__PURE__ */ React__default.createElement("img", {
      src: avatar,
      alt: name,
      className: "lb-avatar-image"
    }), /* @__PURE__ */ React__default.createElement(Initials, null));
  }
);
function getInitials(name) {
  return name.trim().split(" ").reduce((initials, name2, index, array) => {
    if (index === 0 || index === array.length - 1) {
      initials += name2.charAt(0).toLocaleUpperCase();
    }
    return initials;
  }, "");
}

export { Avatar };
