// @liveblocks/react-lexical/dist/mentions/mention-component.js
'use strict';

var LexicalComposerContext = require('@lexical/react/LexicalComposerContext');
var lexical = require('lexical');
var React = require('react');
var index_js = require('use-sync-external-store/shim/index.js');

function Mention({
  nodeKey,
  children
}) {
  const [editor] = LexicalComposerContext.useLexicalComposerContext();
  const isSelected = useIsNodeSelected(nodeKey);
  function handleClick(event) {
    editor.update(() => {
      event.stopPropagation();
      event.preventDefault();
      const selection = lexical.$createNodeSelection();
      selection.add(nodeKey);
      lexical.$setSelection(selection);
    });
  }
  return /* @__PURE__ */ React.createElement("span", {
    onClick: handleClick,
    "data-selected": isSelected ? "" : void 0,
    className: "lb-root lb-lexical-mention"
  }, children);
}
function $isNodeSelected(key) {
  const node = lexical.$getNodeByKey(key);
  if (node === null)
    return false;
  return node.isSelected();
}
function useIsNodeSelected(key) {
  const [editor] = LexicalComposerContext.useLexicalComposerContext();
  const subscribe = React.useCallback(
    (onStoreChange) => {
      return editor.registerUpdateListener(onStoreChange);
    },
    [editor]
  );
  const getSnapshot = React.useCallback(() => {
    return editor.getEditorState().read(() => $isNodeSelected(key));
  }, [editor, key]);
  return index_js.useSyncExternalStore(subscribe, getSnapshot, getSnapshot);
}

exports.Mention = Mention;
