// @liveblocks/react-lexical/dist/mentions/mention-node.mjs
import { createInboxNotificationId } from '@liveblocks/core';
import { DecoratorNode, $applyNodeReplacement } from 'lexical';
import * as React from 'react';
import { Mention } from './mention-component.mjs';
import { User } from './user.mjs';

const MENTION_CHARACTER = "@";
class MentionNode extends DecoratorNode {
  constructor(id, userId, key) {
    super(key);
    this.__id = id;
    this.__userId = userId;
  }
  static getType() {
    return "lb-mention";
  }
  static clone(node) {
    return new MentionNode(node.__id, node.__userId);
  }
  createDOM() {
    const element = document.createElement("span");
    element.style.display = "inline-block";
    element.style.userSelect = "none";
    return element;
  }
  updateDOM() {
    return false;
  }
  static importDom() {
    return {
      span: () => ({
        conversion: (element) => {
          const value = atob(element.getAttribute("data-lexical-lb-mention"));
          const node = $createMentionNode(value);
          return { node };
        },
        priority: 1
      })
    };
  }
  exportDOM() {
    const element = document.createElement("span");
    const value = this.getTextContent();
    element.setAttribute("data-lexical-lb-mention", btoa(value));
    element.textContent = this.getTextContent();
    return { element };
  }
  static importJSON(serializedNode) {
    const node = $createMentionNode(serializedNode.userId);
    return node;
  }
  exportJSON() {
    return {
      userId: this.__userId,
      type: "lb-mention",
      version: 1
    };
  }
  getUserId() {
    const self = this.getLatest();
    return self.__userId;
  }
  getId() {
    const self = this.getLatest();
    return self.__id;
  }
  decorate() {
    return /* @__PURE__ */ React.createElement(Mention, {
      nodeKey: this.getKey()
    }, MENTION_CHARACTER, /* @__PURE__ */ React.createElement(User, {
      userId: this.getUserId()
    }));
  }
}
function $isMentionNode(node) {
  return node instanceof MentionNode;
}
function $createMentionNode(userId) {
  const node = new MentionNode(createInboxNotificationId(), userId);
  return $applyNodeReplacement(node);
}

export { $createMentionNode, $isMentionNode, MentionNode };
