// @liveblocks/react-lexical/dist/mentions/suggestions.mjs
import { useLexicalComposerContext } from '@lexical/react/LexicalComposerContext';
import { KEY_ARROW_DOWN_COMMAND, COMMAND_PRIORITY_LOW, KEY_ARROW_UP_COMMAND, KEY_ESCAPE_COMMAND, KEY_ENTER_COMMAND } from 'lexical';
import React__default, { createContext, forwardRef, useState, useEffect, useRef, useImperativeHandle, useContext } from 'react';

const SuggestionsContext = createContext(null);
const OnValueSelectCallbackContext = createContext(null);
const OnResetMatchCallbackContext = createContext(
  null
);
const HighlightedIndexContext = createContext(null);
const List = forwardRef(
  function(props, forwardedRef) {
    const { children, ...divProps } = props;
    const [editor] = useLexicalComposerContext();
    const [highlightedIndex, setHighlightedIndex] = useState(0);
    const values = useSuggestions();
    const onValueSelect = useOnValueSelectCallback();
    const onEscapeKeyDown = useOnResetMatchCallback();
    useEffect(() => {
      function onKeyArrowDown(event) {
        if (values.length === 0)
          return true;
        if (highlightedIndex === null)
          return true;
        const nextIndex = highlightedIndex === values.length - 1 ? 0 : highlightedIndex + 1;
        setHighlightedIndex(nextIndex);
        event.preventDefault();
        event.stopImmediatePropagation();
        return true;
      }
      return editor.registerCommand(
        KEY_ARROW_DOWN_COMMAND,
        onKeyArrowDown,
        COMMAND_PRIORITY_LOW
      );
    }, [editor, highlightedIndex, values]);
    useEffect(() => {
      function onKeyArrowUp(event) {
        if (values.length === 0)
          return true;
        if (highlightedIndex === null)
          return true;
        const nextIndex = highlightedIndex === 0 ? values.length - 1 : highlightedIndex - 1;
        setHighlightedIndex(nextIndex);
        event.preventDefault();
        event.stopImmediatePropagation();
        return true;
      }
      return editor.registerCommand(
        KEY_ARROW_UP_COMMAND,
        onKeyArrowUp,
        COMMAND_PRIORITY_LOW
      );
    }, [editor, highlightedIndex, values]);
    useEffect(() => {
      function onKeyEscape(event) {
        event.preventDefault();
        event.stopImmediatePropagation();
        onEscapeKeyDown();
        return true;
      }
      return editor.registerCommand(
        KEY_ESCAPE_COMMAND,
        onKeyEscape,
        COMMAND_PRIORITY_LOW
      );
    }, [editor, onEscapeKeyDown]);
    useEffect(() => {
      function onKeyEnter(event) {
        if (values.length === 0)
          return true;
        onValueSelect(values[highlightedIndex]);
        if (event === null)
          return true;
        event.preventDefault();
        event.stopImmediatePropagation();
        return true;
      }
      return editor.registerCommand(
        KEY_ENTER_COMMAND,
        onKeyEnter,
        COMMAND_PRIORITY_LOW
      );
    }, [editor, onValueSelect, highlightedIndex, values]);
    useEffect(() => {
      const root = editor.getRootElement();
      if (root === null)
        return;
      root.setAttribute(
        "aria-activedescendant",
        `typeahead-item-${highlightedIndex}`
      );
      return () => {
        root.removeAttribute("aria-activedescendant");
      };
    }, [editor, highlightedIndex]);
    return /* @__PURE__ */ React__default.createElement(HighlightedIndexContext.Provider, {
      value: [highlightedIndex, setHighlightedIndex]
    }, /* @__PURE__ */ React__default.createElement("div", {
      role: "listbox",
      ...divProps,
      ref: forwardedRef
    }, children));
  }
);
const Item = forwardRef(
  function Item2(props, forwardedRef) {
    const { children, value, onMouseEnter, onClick, ...divProps } = props;
    const divRef = useRef(null);
    const [highlightedIndex, setHighlightedIndex] = useHighlightedIndex();
    const suggestions = useSuggestions();
    const onValueSelect = useOnValueSelectCallback();
    const isHighlighted = suggestions[highlightedIndex] === value;
    useImperativeHandle(
      forwardedRef,
      () => divRef.current
    );
    useEffect(() => {
      if (!isHighlighted)
        return;
      const div = divRef.current;
      if (div === null)
        return;
      div.scrollIntoView({ block: "nearest" });
    }, [isHighlighted]);
    function handleMouseEnter(event) {
      onMouseEnter?.(event);
      if (event.isDefaultPrevented())
        return;
      const index = suggestions.indexOf(value);
      if (index === -1)
        return;
      setHighlightedIndex(index);
    }
    function handleClick(event) {
      onClick?.(event);
      if (event.isDefaultPrevented())
        return;
      onValueSelect(value);
    }
    return /* @__PURE__ */ React__default.createElement("div", {
      role: "option",
      "data-highlighted": isHighlighted || void 0,
      onMouseEnter: handleMouseEnter,
      onClick: handleClick,
      ...divProps,
      ref: divRef
    }, children);
  }
);
function useHighlightedIndex() {
  const context = React__default.useContext(HighlightedIndexContext);
  if (context === null) {
    throw new Error(
      "useHighlightedIndex must be used within a HighlightedIndexProvider"
    );
  }
  return context;
}
function useSuggestions() {
  const suggestions = useContext(SuggestionsContext);
  if (suggestions === null) {
    throw new Error("useSuggestions: SuggestionsContext not found");
  }
  return suggestions;
}
function useOnValueSelectCallback() {
  const onValueSelect = useContext(OnValueSelectCallbackContext);
  if (onValueSelect === null) {
    throw new Error("useOnValueSelectCallback: OnValueSelectContext not found");
  }
  return onValueSelect;
}
function useOnResetMatchCallback() {
  const onResetMatch = useContext(OnResetMatchCallbackContext);
  if (onResetMatch === null) {
    throw new Error("useOnResetMatchCallback: OnResetMatchContext not found");
  }
  return onResetMatch;
}

export { Item, List, OnResetMatchCallbackContext, OnValueSelectCallbackContext, SuggestionsContext };
