// @liveblocks/react-ui/dist/primitives/Composer/contexts.js
'use strict';

var core = require('@liveblocks/core');
var React = require('react');

const ComposerContext = React.createContext(null);
const ComposerEditorContext = React.createContext(null);
const ComposerSuggestionsContext = React.createContext(null);
function useComposerEditorContext() {
  const composerEditorContext = React.useContext(ComposerEditorContext);
  return core.nn(
    composerEditorContext,
    "Composer.Form is missing from the React tree."
  );
}
function useComposer() {
  const composerContext = React.useContext(ComposerContext);
  return core.nn(composerContext, "Composer.Form is missing from the React tree.");
}
function useComposerSuggestionsContext(source = "useComposerSuggestionsContext") {
  const composerSuggestionsContext = React.useContext(ComposerSuggestionsContext);
  return core.nn(
    composerSuggestionsContext,
    `${source} can\u2019t be used outside of Composer.Editor.`
  );
}

exports.ComposerContext = ComposerContext;
exports.ComposerEditorContext = ComposerEditorContext;
exports.ComposerSuggestionsContext = ComposerSuggestionsContext;
exports.useComposer = useComposer;
exports.useComposerEditorContext = useComposerEditorContext;
exports.useComposerSuggestionsContext = useComposerSuggestionsContext;
